flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/SchedulingPipelinedRegionComputeUtil.java

```diff
@@ -0,0 +1,142 @@
+
+        final Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> vertexToRegion =
+                new IdentityHashMap<>();
+
+        // iterate all the vertices which are topologically sorted
+        for (SchedulingExecutionVertex vertex : topologicallySortedVertexes) {
+            Set<SchedulingExecutionVertex> currentRegion = new HashSet<>();
+            currentRegion.add(vertex);
+            vertexToRegion.put(vertex, currentRegion);
+
+            for (ConsumedPartitionGroup consumedPartitionGroup :
+                    vertex.getConsumedPartitionGroups()) {
+                for (IntermediateResultPartitionID consumedPartitionId : consumedPartitionGroup) {
+                    SchedulingResultPartition consumedPartition =
+                            resultPartitionRetriever.apply(consumedPartitionId);
+                    // Similar to the BLOCKING ResultPartitionType, each vertex connected through
+                    // PIPELINED_APPROXIMATE is also considered as a single region. This attribute
+                    // is called  "reconnectable". Reconnectable will be removed after FLINK-19895,
+                    // see also {@link  ResultPartitionType#isReconnectable}
+                    if (!consumedPartition.getResultType().isReconnectable()) {
+                        final SchedulingExecutionVertex producerVertex =
+                                consumedPartition.getProducer();
+                        final Set<SchedulingExecutionVertex> producerRegion =
+                                vertexToRegion.get(producerVertex);
+
+                        // check if it is the same as the producer region, if so skip the merge
+                        // this check can significantly reduce compute complexity in All-to-All
+                        // PIPELINED edge case
+                        if (producerRegion != null && currentRegion != producerRegion) {
+                            currentRegion =
+                                    mergeRegions(currentRegion, producerRegion, vertexToRegion);
+                        }
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+
+        return vertexToRegion;
+    }
+
+    private static List<Integer> buildOutEdgesForRegion(
+            final Set<SchedulingExecutionVertex> currentRegion,
+            final Map<Set<SchedulingExecutionVertex>, Integer> regionIndices,
+            final Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> vertexToRegion,
+            final Function<ExecutionVertexID, ? extends SchedulingExecutionVertex>
+                    executionVertexRetriever) {
+
+        final List<Integer> currentRegionOutEdges = new ArrayList<>();
+        for (SchedulingExecutionVertex vertex : currentRegion) {
+            for (SchedulingResultPartition producedResult : vertex.getProducedResults()) {
+                if (producedResult.getResultType().isPipelined()) {
```

Comment:
Should we make this `!isReconnectable()`?

Reply:
yes of course